In [116]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("erqizhou/students-data-analysis")

print("Path to dataset files:", path)

Path to dataset files: /root/.cache/kagglehub/datasets/erqizhou/students-data-analysis/versions/1


In [117]:
import os

dataset_path = '/root/.cache/kagglehub/datasets/erqizhou/students-data-analysis/versions/1'
files = os.listdir(dataset_path)
print('Nama Files = ', files)

Nama Files =  ['Students data.csv']


In [118]:
import pandas as pd
data_file = os.path.join(dataset_path, 'Students data.csv')
data = pd.read_csv(data_file)

In [119]:
data

ID class  gender  race    GPA  Algebra  Calculus1  Calculus2  \
0    1141     A    male     1  73.47       64         81         87   
1    1142     A  female     1  71.22       57         50         51   
2    1143     A  female     2  74.56       47         48         71   
3    1144     A  female     1  72.89       46         72         38   
4    1145     A  female     1  70.11       49         45         63   
..    ...   ...     ...   ...    ...      ...        ...        ...   
100  1241     A  female     1  88.34       87         83         92   
101  1242     B    male     1  89.84       98         77         95   
102  1243     B    male     1  88.82       83         80         91   
103  1244     A    male     1  86.60       92         82         91   
104  1245     A    male     1  93.71       93         97         99   

     Statistics  Probability  Measure  Functional_analysis from1 from2 from3  \
0            60           74       71                   60     A     A     A   
1            51           55       62                   61     B     A     A   
2            60           61       68                   64     C     A     A   
3            60           29       54                   51     D     A     A   
4            60           66       66                   61     E     A     A   
..          ...          ...      ...                  ...   ...   ...   ...   
100          98           93       86                   90     M     B     A   
101          98           96       88                  100     A     B     A   
102          98           93       95                   71     T     B     A   
103          99           94       82                   78     S     B     A   
104         100           97       90                   90     K     B     A   

     from4  y  
0        3  0  
1        2  0  
2        0  1  
3        0  0  
4        0  0  
..     ... ..  
100      0  1  
101      0  1  
102      0  2  
103      0  2  
104      0  2  

[105 rows x 17 columns]

# <b> Principal component Analysis </b> <br>
Apa itu PCA ? <br>

PCa Merupakan metode statistik yang di gunakan untuk <b> Mereduksi Dimensi</b> dengan cara mengubah data kedalam komponen utama (principal coponent) yang memiliki variasi terbesar, PCA diguanakan untuk mengatasi masalah dimensi padaa data yang memiliki banyak fitur, tanpa kehilangan terlalu banyak informasi penting

<b> Manfaat PCA : </b>
1. eduksi Dimensi
2. Visualisasi Data
3. Mengurangi Overfitting ( menghapus fitur kurang penting )
4. Mempercepat Komputasi

# Perbedaan aakurasi penggunaan 3 data dan 5 data direduksi menjadi  3 dataa menggunakna PCA

In [120]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   105 non-null    int64  
 1   class                105 non-null    object 
 2   gender               105 non-null    object 
 3   race                 105 non-null    int64  
 4   GPA                  105 non-null    float64
 5   Algebra              105 non-null    int64  
 6   Calculus1            105 non-null    int64  
 7   Calculus2            105 non-null    int64  
 8   Statistics           105 non-null    int64  
 9   Probability          105 non-null    int64  
 10  Measure              105 non-null    int64  
 11  Functional_analysis  105 non-null    int64  
 12  from1                105 non-null    object 
 13  from2                105 non-null    object 
 14  from3                105 non-null    object 
 15  from4                105 non-null    int

In [121]:
data.columns

Index(['ID', 'class', 'gender', 'race', 'GPA', 'Algebra', 'Calculus1',
       'Calculus2', 'Statistics', 'Probability', 'Measure',
       'Functional_analysis', 'from1', 'from2', 'from3', 'from4', 'y'],
      dtype='object')

**Klasifikasi menggunakan Support Vector Machine(tanpa PCA)**

In [122]:
x = data[['Algebra', 'Calculus1']]
y = data['y']

In [123]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size= 0.2, random_state= 0)

In [124]:
#support Vector classification
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc = SVC()
model_svc.fit(x_train,y_train)

y_pred = model_svc.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)
print(f'Hasil Akurasi Menggunakan SVM = {akurasi:.2%}')

Hasil Akurasi Menggunakan SVM = 61.90%


**Klasifikasi menggunakan Support Vector Machine(Pakai PCA)**

In [125]:
x = data[['Algebra', 'Calculus1','Calculus2', 'Statistics', 'Probability', 'Measure','Functional_analysis']]
y = data['y']

In [126]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [127]:
#preprocessing
#Normalisasi DAta ( data di - scaling agar memiliki varians yang seragam atau sama)
Scaler = StandardScaler()
x_scaled = Scaler.fit_transform(x)

In [128]:
pca = PCA(n_components = 2)# mengurangi fitur jadi hanya ada 2 fitur
x_pca = pca.fit_transform(x_scaled)

In [129]:
x_pca

array([[-2.07003214e+00,  2.18592202e-02],
       [-5.45839508e+00,  1.05533693e+00],
       [-4.47102836e+00,  1.51839221e+00],
       [-6.79629530e+00, -5.10472372e-01],
       [-4.65746829e+00,  1.20949970e+00],
       [-4.21126236e+00,  8.73626018e-02],
       [-3.71023894e+00,  1.62005742e+00],
       [-4.46008259e+00,  5.18389402e-01],
       [-2.20302007e+00,  8.30426984e-01],
       [-3.51633942e+00, -2.10998039e-03],
       [-3.14513807e+00,  2.90555613e-01],
       [-4.92040566e-01,  1.09154827e+00],
       [-2.02472997e+00, -1.09359314e-01],
       [-1.44634508e+00,  6.07265107e-01],
       [-2.21321917e+00, -2.81474322e-01],
       [-9.32018108e-01,  5.17442152e-01],
       [-2.81760986e+00, -8.50014682e-01],
       [-1.25604790e+00,  8.19180189e-01],
       [-2.44380019e+00, -3.16644381e-01],
       [-1.46377625e+00, -5.78560330e-01],
       [-2.24268431e+00, -5.94289406e-02],
       [ 1.01064578e+00,  1.29459456e-01],
       [-2.00160909e+00, -2.90322643e-01],
       [ 2.

In [130]:
x_train, x_test, y_train, y_test = train_test_split(x_pca, y, test_size= 0.2, random_state= 0)

In [131]:
#support Vector classification
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc = SVC()
model_svc.fit(x_train,y_train)

y_pred = model_svc.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)
print(f'Hsdil Akurasi Menggunakan SVM = {akurasi:.2%}')

Hsdil Akurasi Menggunakan SVM = 66.67%


Daari Hasil pertama Data yang digunakan untuk training hanya menggunakan 2 fitur saja hanayaa memberikan hasil akuraasi sbesar 61.95 saja. Sedaangkan menggunakana Data dengan 5 fitur dan lalu di reduksi menjadi 2 fitur agar informasi dari data ssebelumnya tidak hilang memberikan hasil akurasi sebesaar 66.7% <br>

Kessimpulannya: <br>
Aagar dapat mempertahankan tnpa mengurangi jumlah akurasi yang besar, di sarankan untukmengambil fitur sebanyak baanyaknyaa daan melakukan reduksi menggunakan PCA jika tujuannya untuk Visualisasi 2 dimensi  (2 fitur ) atau 3 dimenssi ( 3 fitur)

# **FEATURE IMPORTNCCE & PCA**

In [132]:
data

ID class  gender  race    GPA  Algebra  Calculus1  Calculus2  \
0    1141     A    male     1  73.47       64         81         87   
1    1142     A  female     1  71.22       57         50         51   
2    1143     A  female     2  74.56       47         48         71   
3    1144     A  female     1  72.89       46         72         38   
4    1145     A  female     1  70.11       49         45         63   
..    ...   ...     ...   ...    ...      ...        ...        ...   
100  1241     A  female     1  88.34       87         83         92   
101  1242     B    male     1  89.84       98         77         95   
102  1243     B    male     1  88.82       83         80         91   
103  1244     A    male     1  86.60       92         82         91   
104  1245     A    male     1  93.71       93         97         99   

     Statistics  Probability  Measure  Functional_analysis from1 from2 from3  \
0            60           74       71                   60     A     A     A   
1            51           55       62                   61     B     A     A   
2            60           61       68                   64     C     A     A   
3            60           29       54                   51     D     A     A   
4            60           66       66                   61     E     A     A   
..          ...          ...      ...                  ...   ...   ...   ...   
100          98           93       86                   90     M     B     A   
101          98           96       88                  100     A     B     A   
102          98           93       95                   71     T     B     A   
103          99           94       82                   78     S     B     A   
104         100           97       90                   90     K     B     A   

     from4  y  
0        3  0  
1        2  0  
2        0  1  
3        0  0  
4        0  0  
..     ... ..  
100      0  1  
101      0  1  
102      0  2  
103      0  2  
104      0  2  

[105 rows x 17 columns]

In [133]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   105 non-null    int64  
 1   class                105 non-null    object 
 2   gender               105 non-null    object 
 3   race                 105 non-null    int64  
 4   GPA                  105 non-null    float64
 5   Algebra              105 non-null    int64  
 6   Calculus1            105 non-null    int64  
 7   Calculus2            105 non-null    int64  
 8   Statistics           105 non-null    int64  
 9   Probability          105 non-null    int64  
 10  Measure              105 non-null    int64  
 11  Functional_analysis  105 non-null    int64  
 12  from1                105 non-null    object 
 13  from2                105 non-null    object 
 14  from3                105 non-null    object 
 15  from4                105 non-null    int

In [142]:
#label Encoding
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
data['class_encoded'] = label_encoder.fit_transform(data['class'])
data['gender_encoded'] = label_encoder.fit_transform(data['gender'])
data['from1_encoded'] = label_encoder.fit_transform(data['from1'])
data['from2_encoded'] = label_encoder.fit_transform(data['from2'])
data['from3_encoded'] = label_encoder.fit_transform(data['from3'])

In [135]:
data

ID class  gender  race    GPA  Algebra  Calculus1  Calculus2  \
0    1141     A    male     1  73.47       64         81         87   
1    1142     A  female     1  71.22       57         50         51   
2    1143     A  female     2  74.56       47         48         71   
3    1144     A  female     1  72.89       46         72         38   
4    1145     A  female     1  70.11       49         45         63   
..    ...   ...     ...   ...    ...      ...        ...        ...   
100  1241     A  female     1  88.34       87         83         92   
101  1242     B    male     1  89.84       98         77         95   
102  1243     B    male     1  88.82       83         80         91   
103  1244     A    male     1  86.60       92         82         91   
104  1245     A    male     1  93.71       93         97         99   

     Statistics  Probability  ...  from1  from2 from3 from4  y  class_encoded  \
0            60           74  ...      A      A     A     3  0              0   
1            51           55  ...      B      A     A     2  0              0   
2            60           61  ...      C      A     A     0  1              0   
3            60           29  ...      D      A     A     0  0              0   
4            60           66  ...      E      A     A     0  0              0   
..          ...          ...  ...    ...    ...   ...   ... ..            ...   
100          98           93  ...      M      B     A     0  1              0   
101          98           96  ...      A      B     A     0  1              1   
102          98           93  ...      T      B     A     0  2              1   
103          99           94  ...      S      B     A     0  2              0   
104         100           97  ...      K      B     A     0  2              0   

     gender_encoded  from1_encoded  from2_encoded  from3_encoded  
0                 1              0              0              0  
1                 0              3              0              0  
2                 0              4              0              0  
3                 0              5              0              0  
4                 0              6              0              0  
..              ...            ...            ...            ...  
100               0             14              1              0  
101               1              0              1              0  
102               1             21              1              0  
103               1             20              1              0  
104               1             12              1              0  

[105 rows x 22 columns]

In [143]:
#menghapus dat kategori
data2 = data.drop(columns = ['ID', 'class', 'gender', 'from1', 'from2', 'from3'])
data2

race    GPA  Algebra  Calculus1  Calculus2  Statistics  Probability  \
0       1  73.47       64         81         87          60           74   
1       1  71.22       57         50         51          51           55   
2       2  74.56       47         48         71          60           61   
3       1  72.89       46         72         38          60           29   
4       1  70.11       49         45         63          60           66   
..    ...    ...      ...        ...        ...         ...          ...   
100     1  88.34       87         83         92          98           93   
101     1  89.84       98         77         95          98           96   
102     1  88.82       83         80         91          98           93   
103     1  86.60       92         82         91          99           94   
104     1  93.71       93         97         99         100           97   

     Measure  Functional_analysis  from4  y  class_encoded  gender_encoded  \
0         71                   60      3  0              0               1   
1         62                   61      2  0              0               0   
2         68                   64      0  1              0               0   
3         54                   51      0  0              0               0   
4         66                   61      0  0              0               0   
..       ...                  ...    ... ..            ...             ...   
100       86                   90      0  1              0               0   
101       88                  100      0  1              1               1   
102       95                   71      0  2              1               1   
103       82                   78      0  2              0               1   
104       90                   90      0  2              0               1   

     from1_encoded  from2_encoded  from3_encoded  
0                0              0              0  
1                3              0              0  
2                4              0              0  
3                5              0              0  
4                6              0              0  
..             ...            ...            ...  
100             14              1              0  
101              0              1              0  
102             21              1              0  
103             20              1              0  
104             12              1              0  

[105 rows x 16 columns]

In [144]:
#melakukan analisis data murid yang diperlukan
x = data2.drop(columns = ['y'])
y = data['y']

In [145]:
#support vektor classification
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc2 = SVC(kernel = 'linear')
model_svc2.fit(x_train, y_train)

y_pred = model_svc2.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)
print(f'Hasil Akurasi menggunakan SVM  = {akurasi:.2%}')

Hasil Akurasi menggunakan SVM  = 38.10%


In [147]:
import numpy as np
feature_importance = pd.DataFrame({
    'Feature': x.columns,
    'Importance': np.abs(model_svc.coef_[0]) #Khusus Klasifikasi, pakai absolute
})

feature_importance = feature_importance.sort_values(by='Importance', ascending=False)
print(feature_importance)

ValueError: All arrays must be of the same length

Berdaasarkan Feature importance, terdapat 8 fitur  yang nilainya kurang dari 0,1 sehinggaa fitur tersebut tidak memberikan dampak besar pada model yang digunakan, jadi perlu dilakukan training ulang denhan mereduksi dimensi menjadi 6 fitur, kerana memberikan nilai lebih dari0,1

In [148]:
#Preprocessing
#normalisasi data (data di-scaling agar memiliki varians yang seragam)
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

In [149]:
pca = PCA(n_components = 6) #ini mengurangi fitur menjadi 6 fitur
x_pca = pca.fit_transform(x_scaled)

In [150]:
x_train, x_test, y_train, y_test = train_test_split(x_pca, y, test_size=0.2, random_state = 0)

In [151]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model_svc2 = SVC(kernel='linear')
model_svc2.fit(x_train, y_train)

y_pred = model_svc2.predict(x_test)
akurasi = accuracy_score(y_test, y_pred)
print(f'hasil akurasi menggubakan svm = {akurasi:.2%}')

hasil akurasi menggubakan svm = 57.14%


Kesimpulan : <br>
Ternayata